### ISSUE TITLE
Timeout when add to group

### ISSUE THREAD

**Issue by TresaL** [Azure/azure-signalr#745]
Hi, 
We have a Web API (hosted in azure using app service plan) configured with ASRS. At times, we are experiencing timeouts when add connection into a group. Here is our hub method to add to group.

public async Task AddToGroup(string groupName)
        {
            await Groups.AddToGroupAsync(Context.ConnectionId, groupName);
        }

Here is the exception message 
"Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out." and trace
System.TimeoutException:
   at Microsoft.Azure.SignalR.ServiceConnectionContainerBase+<WriteAckableMessageAsync>d__49.MoveNext (Microsoft.Azure.SignalR.Common, Version=1.0.14.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.SignalR.MultiEndpointServiceConnectionContainer+<WriteAckableMessageAsync>d__18.MoveNext (Microsoft.Azure.SignalR.Common, Version=1.0.14.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at BDL.TwilioAPI.Hub.CallStatusHub+<AddToGroup>d__0.MoveNext (TwilioAPI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=nullTwilioAPI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: D:\a\1\s\TwilioAPI\Hub\CallStatusHub.csTwilioAPI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: 9)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1+<ExecuteHubMethod>d__15.MoveNext (Microsoft.AspNetCore.SignalR.Core, Version=1.0.4.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1+<Invoke>d__13.MoveNext (Microsoft.AspNetCore.SignalR.Core, Version=1.0.4.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)

What would be the cause and how can we fix this? Our clients facing much of inconvenience in retrying. The ASRS is configured with "Classic" mode and Standard pricing tier. The ASRS metrics were alright in terms  on server and client connections when these timeout occurs.

We are using azure-signalr 1.0.14 version.

What would be the cause for this issue and is there any workaround that can help us?

Regards,
Tresa

**Comment by JialinXin** [comment]
@TresaL Thanks for reporting the issue. Would you send your resourceId and time when you met the issue to _jixin[at]microsoft.com_. We'll look into this and reply you soon.

**Comment by JialinXin** [comment]
@TresaL I haven't got your resource information yet. But looking into one of our cases similar to your case. This might be related to client connection disconnect near the join group message which leads to ack not able to handle correctly. It won't affect live clients and we'll improve the server side logic to handle this kind of case. #750 
You can check server side traces of connectionId between the exception ones and disconnect ones to see if they are matched. And email me your resource details it doesn't your cases.

**Comment by TresaL** [comment]
Thanks, i’ve forwarded you resource ids. Please let me know if you don’t receive them.  I’m pretty sure the client connection was active during timeout. I noticed the client ping message traces during this error.

**Comment by JialinXin** [comment]
@TresaL Sorry for the delay response. I've replied you details in the email. The complete fix will be releasing soon. For short workaround, please try **restart** the service.

**Comment by ajbeaven** [comment]
Were any fixes pushed out for this? I'm also hitting the same exception every so often:
```
System.TimeoutException: Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out.
  Module "Microsoft.Azure.SignalR.ServiceConnectionContainerBase", in WriteAckableMessageAsync
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  Module "Microsoft.Azure.SignalR.MultiEndpointServiceConnectionContainer+<>c__DisplayClass24_0+<<WriteAckableMessageAsync>b__0>d", in MoveNext
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  Module "Microsoft.Azure.SignalR.MultiEndpointServiceConnectionContainer+<>c__DisplayClass28_0+<<WriteMultiEndpointMessageAsync>b__2>d", in MoveNext
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  Module "Microsoft.Azure.SignalR.MultiEndpointServiceConnectionContainer", in WriteAckableMessageAsync
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  Module "Microsoft.Azure.SignalR.AspNet.ServiceMessageBus", in WriteMessage
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  File "C:\agent\_work\7\s\ChatR\Services\ChatService.cs", line 1485, col 5, in EnsureChatClientPresentAsync
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  File "C:\agent\_work\7\s\ChatR\Services\ChatService.cs", line 127, col 4, in NotifyServerThatClientIsReadyAsync
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  File "C:\agent\_work\7\s\ChatR\Hubs\ChatHub.cs", line 57, col 4, in NotifyServerThatClientIsReady
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  Module "Microsoft.AspNet.SignalR.Hubs.HubPipelineModule+<>c__DisplayClass0_0+<<BuildIncoming>b__0>d", in MoveNext
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  Module "Microsoft.AspNet.SignalR.Hubs.HubPipelineModule+<>c__DisplayClass0_0+<<BuildIncoming>b__0>d", in MoveNext
  Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
  Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
  Module "Microsoft.AspNet.SignalR.Hubs.HubPipelineModule+<>c__DisplayClass0_0+<<BuildIncoming>b__0>d", in MoveNext
```

Unlike OP, I'm running ASRS with Microsoft.Azure.SignalR.AspNet (1.4.3) on .NET Framework (4.7.2).

**Comment by JialinXin** [comment]
@ajbeaven The fix for original issue is done. The reason for this kind of exception can be various. Would you share us(_jixin[at]microsoft.com_) further information about time, resourceId about your case?

**Comment by ajbeaven** [comment]
Just adding a note that upon investigation, it appears that my specific case was related to #750 and should be resolved when a fix for that has been pushed out.

**Comment by mfarnz** [comment]
I'm also seeing this same timeout when JoinGroupWithAckMessage is called. Seems to happen when there is lots of traffic on our hub server. I've tried adding more app server instances and scaling up ASRS. But no luck. We're using Microsoft.Azure.SignalR 1.5.1 and .NET Core 3.1. Any help or guidance on debugging would be greatly appreciated.

**Comment by JialinXin** [comment]
@mfarnz Usually this kind of issue could be caused by 1. App server too busy to handle this requests, 2. Clients quickly leaves while join groups already called. Case 2 should already be mitigated in service side. Suggest you enable server-side logs(refer to [Link](https://docs.microsoft.com/en-us/azure/azure-signalr/signalr-howto-troubleshoot-method#enable-server-side-logging-for-aspnet-core-signalr)) to evaluate the times of connections connect, disconnect, join group, and time-out. Server side will set timeout if fail to receive ack within 10 seconds. Possibly you may take a look if it's the case.

**Comment by mfarnz** [comment]
@JialinXin in the app-server too busy case, can adjusting the number of service connections help? Not sure if that effects who many Ack message that can be handled simultaneously. Another thing to note, we're calling the join group (along with some db calls) from our hub's OnConnectedAsync handler. Could this be problematic?

**Comment by JialinXin** [comment]
@mfarnz How many service connection are you configured now? 5 is the default value should be enough in most cases, and suggest to reduce no less than 2 for a better robustness. As our doc suggested, if there're too much clients, it's suggested to increase the service connection a little for a better throughput. 

You may take a look at server side CPU/Memory which could be a signal of whether server is too busy to handle the requests. And may further look up which components are taking the traffic. 

> This option controls the count of connections between application server and Azure SignalR Service. It is not recommended to change the default value since it should be enough in most of the time. You can increase it for better performance if the total client count is too big. For example, if you have 100,000 clients in total, the connection count can be increased to 10 or 15 for better throughput.

**Comment by mfarnz** [comment]
Currently, we're using the default service connection count. It's worth noting that even though our connection count is not particularly high, there easily be 100 or more groups at any given time. Seems like that should be fine, though.
________________________________

**Comment by brechtvhb** [comment]
I am also receiving this error. Even with no load on the server.

Here's the stack trace:

System.TimeoutException: Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out.
at Microsoft.Azure.SignalR.ServiceConnectionContainerBase.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass8_0.<<WriteAckableMessageAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass9_0.<<WriteMultiEndpointMessageAsync>b__2>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.ServiceLifetimeManagerBase`1.WriteCoreAsync[T](T message, Func`2 task)

**Comment by laszlozold** [comment]
I started seeing these errors right after I switched my solution from dotnet core 3.1 to .net 5. I also upgraded Microsoft.Azure.SignalR from 1.6.1 to 1.8.0. Turned out, when I downgrade the package version back to 1.6.1 it starts working again. I did not try any version between the two though. So my solution was to go with 1.6.1 as it works fine.

**Comment by brechtvhb** [comment]
@laszlozold Oh my god! The issue is indeed fixed by reverting to 1.6.1.

@JialinXin Is this a known issue at microsoft and is there being worked on a fix?
I reported the issue nearly 3 months ago as a seperate bug but nothing happens with it:
https://github.com/Azure/azure-signalr/issues/1225

Some communication about this issue would be really welcome.

**Comment by zackliu** [comment]
@brechtvhb @laszlozold Is there any special usage when you're meeting the exception? I mean does the `Groups.AddToGroupAsync` in a `OnConnectedAsync` or just in normal method? Are there any other methods like send to group or broadcast together with join group operations? What's the frequency you're meeting the timeout exception? After you downgrade to 1.6.1, the exception is totally disappeared or just the frequency becomes lower?
I'm asking these questions because it's hard to reproduce from my side. I used an endless while loop and called `AddToGroupAsync` and `RemoveFromGroupAsync`, but no exception got thrown.

**Comment by brechtvhb** [comment]
@zackliu 

I am calling the AddToGroupAsync from the OnConnectedAsync method, other than that I don't do anything related to signalR in that method. I tried with and without cancellation token but I do receive the error on both cases.
I do an (async) database call to fetch a user object from the database right before calling the AddToGroupAsync method.

Most users are part of multiple (on average 6) groups so I do add them to a group within a foreach.

```
            foreach (var group in groups)
            {
                await Groups.AddToGroupAsync(Context.ConnectionId, group.Id);
                _logger.LogInformation($"User was added to SignalR group {group.Id}");
            }
```

I received the exception about 80% of the time. I haven't seen the exception at al since reverting to 1.6.1.
If needed I can try to reproduce the bug in a test repo.

**Comment by zackliu** [comment]
The frequency is unbelievable high. A simple test code is great for us to investigate. I tried in a relative high load but didn't reproduce.

```
public override async Task OnConnectedAsync()
{
    await Task.WhenAll(Enumerable.Range(1, 100).Select(_ => Test()).ToList());
}

private async Task Test()
{
    var groupName = Guid.NewGuid().ToString();

    while (true)
    {
        try
        {
            var v = Interlocked.Increment(ref _index);
            await Groups.AddToGroupAsync(Context.ConnectionId, groupName);
            for (int i = 0; i < 10; i++)
            {
                var g = Guid.NewGuid().ToString();
                await Groups.AddToGroupAsync(Context.ConnectionId, g);
            }
            await Clients.Group(groupName).SendAsync("broadcastMessage", v.ToString(), DateTime.Now);
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, groupName);
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex);
        }
    }
}
```

**Comment by brechtvhb** [comment]
Here's an isolated test case that causes the error on 1.6.1 and not on 1.8.1

```
    [Authorize]
    public class NominationsHub : Hub
    {
        private readonly ILogger<NominationsHub> _logger;
        public NominationsHub(ILogger<NominationsHub> logger)
        {
            _logger = logger;
        }
        public override async Task OnConnectedAsync()
        {

            await AddToSignalRGroups();
            await base.OnConnectedAsync();
        }
        public override async Task OnDisconnectedAsync(Exception exception)
        {
            await RemoveFromSignalRGroups();
            await base.OnDisconnectedAsync(exception);
        }


        private async Task AddToSignalRGroups()
        {
            await Groups.AddToGroupAsync(Context.ConnectionId, "TERMINAL1");
            await Groups.AddToGroupAsync(Context.ConnectionId, "TERMINAL2");
            await Groups.AddToGroupAsync(Context.ConnectionId, "TERMINAL3");
            await Groups.AddToGroupAsync(Context.ConnectionId, "TERMINAL4");
            await Groups.AddToGroupAsync(Context.ConnectionId, "TERMINAL5");
            await Groups.AddToGroupAsync(Context.ConnectionId, "TERMINAL6");
            await Groups.AddToGroupAsync(Context.ConnectionId, "TERMINAL7");

            _logger.LogInformation($"User {Context.ConnectionId} was added to SignalR terminals");
        }

        private async Task RemoveFromSignalRGroups()
        {
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, "TERMINAL1");
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, "TERMINAL2");
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, "TERMINAL3");
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, "TERMINAL4");
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, "TERMINAL5");
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, "TERMINAL6");
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, "TERMINAL7");

            _logger.LogInformation($"User {Context.ConnectionId} was removed from SignalR terminals.");
        }
    }
```

Error:
```
[2021-05-07 10:18:35Z WRN] Failed to send message null.
System.TimeoutException: Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out.
at Microsoft.Azure.SignalR.ServiceConnectionContainerBase.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass8_0.<<WriteAckableMessageAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass9_0.<<WriteMultiEndpointMessageAsync>b__2>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.ServiceLifetimeManagerBase`1.WriteAckableCoreAsync[T](T message, Func`2 task)
[2021-05-07 10:18:35Z ERR] Error when dispatching 'OnConnectedAsync' on hub.
System.TimeoutException: Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out.
at Microsoft.Azure.SignalR.ServiceConnectionContainerBase.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass8_0.<<WriteAckableMessageAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass9_0.<<WriteMultiEndpointMessageAsync>b__2>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.ServiceLifetimeManagerBase`1.WriteAckableCoreAsync[T](T message, Func`2 task)
at SI.Nominations.Web.Api.Hubs.NominationsHub.AddToSignalRGroups(User user) in /src/SI.Nominations.Web.Api/Hubs/NominationsHub.cs:line 33
at SI.Nominations.Web.Api.Hubs.NominationsHub.OnConnectedAsync() in /src/SI.Nominations.Web.Api/Hubs/NominationsHub.cs:line 21
at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.OnConnectedAsync(HubConnectionContext connection)
at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.OnConnectedAsync(HubConnectionContext connection)
at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.RunHubAsync(HubConnectionContext connection)
```

Base docker image is mcr.microsoft.com/dotnet/aspnet:5.0-alpine3.12.
Running on AKS in EU-West

**Comment by zackliu** [comment]
@brechtvhb It's very weird that I can't reproduce the exception in neither 1.8.1 nor 1.6.1. I checked some code changes in SDK between 1.6.1 and 1.8.1, could you have a test in both 1.7.0 and 1.7.1 and verify whether they will throw exceptions. Also, please give provide the ResourceId and the time you met exceptions so that we can check more logs in service side and find out what was happening.

**Comment by brechtvhb** [comment]
@zackliu Tomorrow I have time to test with 1.7.0 and 1.7.1. Will send you the resource ID after that so I can provide you with a timeframe of where we had failures.

**Comment by brechtvhb** [comment]
Hi. I don't seem to have the issue in 1.7.0. I do have the issue starting from 1.7.1.
I sent you the resource ID by e-mail.

**Comment by gjdonkers** [comment]
@zackliu I'm having the same issue. In 1.8.1. I'm seeing the same issue as @brechtvhb but when downgrading to 1.6.1 the error is gone. If you need more details please let me know.

**Comment by KKhurin** [comment]
This (and #1225) seems to be not reproducible in every environment.
@gjdonkers, @brechtvhb, would you be willing to run an instrumented private (with more logs and cross checks) and share the logs once you repro it?

**Comment by gjdonkers** [comment]
@KKhurin I am willing to do that. At what e-mailadres can I contact you the best?

**Comment by KKhurin** [comment]
> @KKhurin I am willing to do that. At what e-mailadres can I contact you the best?

Thank you for offering help with figuring this out. You can contact me at <my github alias>@microsoft.com

**Comment by brechtvhb** [comment]
I am available for help.
Already had contact @zackliu about this too.

**Comment by KKhurin** [comment]
Unfortunately we still can not repro this locally and the efforts to instrument the code so the logs would contain enough information about the culprit have stalled due to having too many possible areas to add the logs.
To narrow down possible area of instrumentation, could anyone please try running the code that adds to the group in the following wrapper and report if you can still repro the timeouts?
~~~
using (ExecutionContext.SuppressFlow())
{
    Task.Run(FuncCallingAddToGroup);
}
~~~

**Comment by surexxx** [comment]
@KKhurin using your wrapper I get this error: System.InvalidOperationException: 'AsyncFlowControl object must be used on the thread where it was created.'
![image](https://user-images.githubusercontent.com/18722513/137892410-4f21981f-c9b8-4c43-b838-a2d39d5a3781.png)

**Comment by KKhurin** [comment]
Re: `"AsyncFlowControl object must be used..."` - yeah, you'd have to avoid using await inside `SuppressFlow` and perhaps use some other mechanism (e.g. TaskCompletionSource) to return back the task from `AddToGroupAsync` to the original caller.

**Comment by robertlyson** [comment]
I'm experiencing the same issue (`waiting for ack timed out`) coming from my overridden `OnConnectedAsync`. At first, I thought it's because of using the free tier but upgrading to the standard didn't solve the issue.

```
public override async Task OnConnectedAsync()
{
    var user = Context.User?.ToUser();
    if (user == null)
    {
        return;
    }
    var tenantIdAsync = await TenantIdAsync();
    var groups = user.SignalrGroup(tenantIdAsync).Distinct().ToArray();
    var userId = await UserIdAsync(user);
    foreach (var @group in groups)
    {
        _logger.LogInformation("User {Id} ({ConnectionId}) added to group {Group}", userId, Context.ConnectionId, group);
        await _context.Groups.AddToGroupAsync(Context.ConnectionId, group, Context.ConnectionAborted);
    }
    await base.OnConnectedAsync();
}
```

Downgraded `Microsoft.Azure.SignalR` from `1.12.0` to `1.6.1` and issue is gone.

**Comment by vicancy** [comment]
I am trying an approach to fix this issue according to the assumption from @KKhurin. 
Could anyone try preview package `--version 1.13.0-preview1-10900` 
```
 dotnet add package Microsoft.Azure.SignalR --version 1.13.0-preview1-10900 --source https://www.myget.org/F/azure-signalr-dev/api/v3/index.json
```
 to see if it mitigates the issue?

**Comment by robertlyson** [comment]
@vicancy I just gave it a try and seems like the issue is gone in `1.13.0-preview1-10900` will give it more testing tomorrow, thanks for looking into this, great work 👍

**Comment by vicancy** [comment]
Keep it open for now.

**Comment by vicancy** [comment]
We released 1.13.0 containing the mitigation.

**Comment by brechtvhb** [comment]
Fix tested & confirmed. Thank you!

**Comment by KKhurin** [comment]
Folks who ran into this, could you please confirm that in all these cases you'd have to have multiple service endpoints to be able to repro the timeouts? We have found and fixed the culprit for them (see #1576), just wanted to confirm that there are no other unaddressed issues.

**Comment by ajbeaven** [comment]
I was never able to reliable reproduce this in my environment and it was pretty rare (last received this 1 month ago). This seems to support the idea this is caused by a race condition. Yes we do have multiple service endpoints.
